Add MPI cluster resolver and update documentation of SLURM cluster resolver

Flamefire opened the pull request:
@frankchn As discussed I updated the documentation for the Slurm cluster resolver enhanced in #36159 which became outdated.

I also added the MPI cluster resolver mentioned there and included it in the documentation update to avoid conflicts or dependencies between those 2 related changes.

2 design decisions of the MPI cluster resolver I wanted to highlight:
- import of mpi4py only on constructor (only place where it is needed) to allow importing the file even when the package is not installed. Can be useful when conditionally switching on the used resolver and can reduce program startup time.
- removed parameter `tasks_per_node` (actually just didn't add it). As I highlight in the Slurm doc file using it is usually not a good idea as the resolver is able to do that itself and it is very easy to use wrong and hard to use right. I don't see a usecase for that, especially when using MPI

Flamefire committed:
Add MPIClusterResolver for running in generic MPI jobs

This allows usage of programs launched by mpirun or (in extension)
pretty much any batch system.
mpi4py is required to instantiate the cluster resolver

Flamefire committed:
Fix minor documentation issue regarding used Slurm variables

Flamefire committed:
Update documentation of Slurm cluster resolver including MPI

Removes the outdated documentation file and create a proper one with up
to date description and usage examples.
Also use .md as the extension for proper display in editors and on Github.

frankchn commented on tensorflow/python/distribute/cluster_resolver/mpi_cluster_resolver.py:
```diff
@@ -0,0 +1,97 @@
+      RuntimeError: If requested more GPUs per node then available or
+        requested more tasks then assigned tasks.
+    """
+    from mpi4py import MPI # pylint: disable=import-outside-toplevel
```
Maybe catch the ImportError and return a better error message directing users to install mpi4py? Something like https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/distribute/cluster_resolver/tpu_cluster_resolver.py#L32-L34 works.

Flamefire replied:
would a `raise ImportError("mpi4py could not be imported. Run 'pip install mpi4py' to install")` work for you? Logging isn't enough in this case.

frankchn replied:
Yah works for me!

Flamefire replied:
Done in 6a47a283b0

Flamefire replied:
Sorry hat a formatting issue and had to force-push the last commit. Strange though

frankchn reviewed (changes requested):
Just one change from me. Thanks for writing this up!

Flamefire committed:
Show pip install help when failing to import mpi4py

Flamefire commented:
@frankchn The error is
> do_check_futures_test: Check that python files have certain __future__ imports

I omitted those present in other files as they are only required for Python 2 (AFAIK) and TF announced dropping Python 2 after 2.1(?) Isn't that true? Should the check be changed or shall I add the future imports? That would be
```
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function
```

Although nothing of that is used...

frankchn commented:
@Flamefire For consistency can you just add it in? I think all our files have it since we have some code internally still on Py2 so we need to maintain compatibility right now. Thanks!

Flamefire committed:
Add future imports for Python 2 compatibility

Flamefire commented:
>  I think all our files have it since we have some code internally still on Py2 so we need to maintain compatibility right now

Oh, so what is about that announcement:

> TensorFlow will also stop supporting Python 2 starting January 1, 2020

Do I also have to adapt the `super()` call to the Python 2 version? Added the future imports for now to make CI happy.

frankchn commented:
Yeah for the open source builds we are not supporting Python 2 any more, but internally within Google the migration to Python 3 isn't complete, so that's why the pre-submit checks are still around to enforce the compatibility headers.

Your code won't be run internally, so I think it would be fine to leave it as is otherwise.

Flamefire commented:
I see, thanks for the explanation. Then this is ready (again) I believe

Flamefire commented:
@frankchn Seeing a (possibly unrelated) failure:
> //tensorflow/compiler/mlir/tensorflow/tests/mlir2graphdef:tf_add.mlir.test FAILED

Is this good to go anyway? Maybe even for 2.2 as https://github.com/tensorflow/tensorflow/pull/36159 is part of 2.2 and this is the corresponding docu update.

Maybe one final thing: There is a question https://github.com/tensorflow/tensorflow/issues/36094#issuecomment-609973779 about how to use this and indeed there is no documentation on how to use a ClusterResolver with a Strategy in an example. I could add it here, but I think it would be even better to add it to the main docu. So probably out of scope of this PR.

frankchn commented:
Yeah that test failure is unrelated. 

I think we are getting some internal tests done and this will be merged in the next day or two. Not sure if this will make it into 2.2 though since there has already been RC cuts and they are only really willing to accept bug fixes at this point. Will definitely be in 2.3 (cut in ~May-ish) though.

frankchn commented:
Hi @Flamefire, just got updated on an internal discussion and we are actually thinking of moving some of the cluster resolvers out of the TPU repo into a repo owned by a SIG instead (maybe SIG-Networking). 

Still happy to accept the documentation update though, so can you remove the mpi_cluster_resolver.py from the PR, and create a new Issue re: mpi_cluster_resolver and assign it to me and @jhseu? 

We'll update you on whether we decide to move the cluster resolvers soon. Thanks!

frankchn commented:
Sorry about the back and forth!

Flamefire commented:
Moved the documentation update to a new PR with an appropriate branch and updated commit and readme. --> https://github.com/tensorflow/tensorflow/pull/38355

Issue Re MPI resolver added: https://github.com/tensorflow/tensorflow/issues/38356

Closing this.
